### ISSUE TITLE
RewardedAdEvent.earnedReward never gets called on iOS

### ISSUE THREAD

**Issue by fly-qp** [bdlukaa/native_admob_flutter#40]
I tested multiple times but on reward is never ben called when showing a rewarded ad on iOS!
Please fix.

**Comment by bdlukaa** [comment]
Hey, thanks for reporting this issue! Does this happen on Android too? Are you using `Rewarded Ads` or `Rewarded Interstitial Ads`?

**Comment by fly-qp** [comment]
Reward Ads - no not on android

**Comment by bdlukaa** [comment]
Hey! I could not reproduce this issue. What version of the plugin are you using?

**Comment by fly-qp** [comment]
native_admob_flutter: ^1.2.0

**Comment by bdlukaa** [comment]
I can confirm this is also happening on Android with the latest version!

**Comment by bdlukaa** [comment]
Okay I managed to fix it. Nothing wrong on the platform side. It was just a casting error on the dart side. I'll release a new version with the fix for it soon

**Comment by fly-qp** [comment]
Thanks man I appreciate it very much 😍💕
Please let me know when new version is out

**Comment by bdlukaa** [comment]
Published [v1.2.1](https://pub.dev/packages/native_admob_flutter) with the fix for it.

**Comment by niypoo** [comment]
> Published [v1.2.1](https://pub.dev/packages/native_admob_flutter) with the fix for it.

I'm facing same issue now , I'm using last version `native_admob_flutter: ^1.2.4`

**Comment by Billy-80** [comment]
@bdlukaa i am facing this issue on IOS with the test reward ad, same exact code works ok on Android. RewardedAdEvent.closed is called without issues but case RewardedAdEvent.earnedReward is not called.

**Comment by bdlukaa** [comment]
ooh wow. I'll investigate

**Comment by bdlukaa** [comment]
I don't know what's wrong on the iOS side. It seems to be everything good.

Swift code:
```swift
            rewardedAd.present(fromRootViewController: (UIApplication.shared.keyWindow?.rootViewController)!) { () in
                self.channel.invokeMethod("onUserEarnedReward", arguments: [
                    "amount": self.rewardedAd.adReward.amount,
                    "type": self.rewardedAd.adReward.type,
                ])
            }
```

https://developers.google.com/admob/ios/rewarded-ads#display_the_ad_and_handle_the_reward_event

Kotlin code:

```kotlin
                rewardedAd!!.show(activity) { reward ->
                    channel.invokeMethod("onUserEarnedReward", hashMapOf<String, Any>(
                            "amount" to reward.amount,
                            "type" to reward.type
                    ))
                }
```

Dart code:
```dart
      case 'onUserEarnedReward':
        onEventController.add({
          RewardedAdEvent.earnedReward: RewardItem.fromJson(call.arguments)
        });
        break;
```

Anyone with experience knows what's wrong?

**Comment by niypoo** [comment]
What was the wrong at first time you have known about this issue?
I think at version 1.2.1

**Comment by bdlukaa** [comment]
> What was the wrong at first time you have known about this issue?

The data parsing from `RewardItem.fromJson(call.arguments)`. It was not working on Android too, but I fixed it an it's now working on Android. Idk why it doesn't work on iOS since the implementation is basically the same.

I can't test it on iOS since all the iOS implementation is made from the community. It'd be very helpful someone to help me out.

**Comment by niypoo** [comment]
So , can you disabled any data return like data of award , just trigger a event and see if the event trigger or not

**Comment by bdlukaa** [comment]
The fix for it may be https://github.com/vishwajit76/native_admob_flutter/commit/96189d8a4214e52dd2eced1543256d028b1735cd, can you try his fork? Still, I can't test it because I don't have an Apple env.

**Comment by niypoo** [comment]
> The fix for it may be [vishwajit76@96189d8](https://github.com/vishwajit76/native_admob_flutter/commit/96189d8a4214e52dd2eced1543256d028b1735cd), can you try his fork? Still, I can't test it because I don't have an Apple env.

okay , give me a full fork repo url.

**Comment by bdlukaa** [comment]
https://github.com/vishwajit76/native_admob_flutter

**Comment by niypoo** [comment]
> https://github.com/vishwajit76/native_admob_flutter

I can't test this fork,
it's gives to me when try initialize `MobileAds.initialize`

```
Unhandled Exception: 'package:native_admob_flutter/src/utils.dart': Failed assertion: line 23 pos 10: 'Platform.isAndroid': Android is the only supported platform for now
```

and also the initialize function refuse this argument `rewardedInterstitialAdUnitId ` 
```
The named parameter 'rewardedInterstitialAdUnitId' isn't defined.
Try correcting the name to an existing named parameter's name, or defining a named parameter with the name 'rewardedInterstitialAdUnitId'.
```

**Comment by bdlukaa** [comment]
I think I know where this error is coming from. [GADAdReward#amount](https://developers.google.com/admob/ios/api/reference/Classes/GADAdReward.html#amount) is NSDecimalNumber. It's probably just a parsing issue. I'll keep investigating

**Comment by bdlukaa** [comment]
https://stackoverflow.com/a/28268951/11622876 should be the fix for it.

**Comment by zenkog** [comment]
Okay so how do I solve it in my Flutter code since this link is swift? 

> https://stackoverflow.com/a/28268951/11622876 should be the fix for it.
> 
> https://flutter.dev/docs/development/platform-integration/platform-channels#codec

**Comment by bdlukaa** [comment]
Can you guys try the latest master and see if it works?

```yaml
dependencies:
  native_admob_flutter:
    git:
      url: https://github.com/bdlukaa/native_admob_flutter.git
```

**Comment by niypoo** [comment]
> Can you guys try the latest master and see if it works?
> 
> 
> 
> ```yaml
> 
> dependencies:
> 
>   native_admob_flutter:
> 
>     git:
> 
>       url: https://github.com/bdlukaa/native_admob_flutter.git
> 
> ```

Okay I will try it today, thanks

**Comment by zenkog** [comment]
> Can you guys try the latest master and see if it works?
> 
> ```yaml
> dependencies:
>   native_admob_flutter:
>     git:
>       url: https://github.com/bdlukaa/native_admob_flutter.git
> ```

I tried just now but it is still doesn't work. 

I still go from `loaded`, then I watch, then it fire `showed` then once I close it only fire `closed`.

```
flutter: *** Ad Loaded!
flutter: *** Ad showed
flutter: *** Ad closed
```

**Comment by zenkog** [comment]
Hi. Sorry don't mean to push, but any update? I'm behind on a deadline.

**Comment by zenkog** [comment]
And if I understand the behavior correctly. In Android, when I finished watching a video Ad and then I close it, it fires `rewarded` first before firing `close`. Is it supposed to be the same behavior in iOS ?

**Comment by niypoo** [comment]
> And if I understand the behavior correctly. In Android, when I finished watching a video Ad and then I close it, it fires `rewarded` first before firing `close`. Is it supposed to be the same behavior in iOS ?

Yes its a correct behavior, rewarded trigger before close event .

**Comment by bdlukaa** [comment]
Sorry but I won't be able to work on this today, only tomorrow!

**Comment by zenkog** [comment]
No Problem. Thanks for the reply. 

> Sorry but I won't be able to work on this today, only tomorrow!

**Comment by bdlukaa** [comment]
Please see https://github.com/bdlukaa/native_admob_flutter/issues/58

**Comment by niypoo** [comment]
Any update !! In this issue?

**Comment by genox-wang** [comment]
try debugged onUserEarnedReward  event
```dart
 case 'onUserEarnedReward':
      try {
        onEventController.add({
          RewardedAdEvent.earnedReward: RewardItem.fromJson(call.arguments)
        });
      } catch(e) {
        print(e);
      }
```
I found `amount` field is double, but try to parse it with int

![WX20210603-120109](https://user-images.githubusercontent.com/6120793/120585724-ddafed00-c464-11eb-9579-31534e9a5557.png)
![WX20210603-120729](https://user-images.githubusercontent.com/6120793/120585727-dee11a00-c464-11eb-89cf-3eb2beb86106.png)

**Comment by bdlukaa** [comment]
@genox-wang thanks for your info. It'll be very helpful

**Comment by flutter2k20** [comment]
> @genox-wang thanks for your info. It'll be very helpful

Any plans on including the fix from @genox-wang in the next release?

**Comment by shofizone** [comment]
Any update on this issue? I got a bad review for this.

**Comment by flutter2k20** [comment]
Is there any update to this issue?

**Comment by bdlukaa** [comment]
Published version 1.4.1 that is supposed to fix this issue. Thanks @zhengbomo for the fix for it #85
